### PR TITLE
Feature multiselect notes delete and move to another folder

### PR DESCRIPTION
**Pull request by voidsatisfaction**
# WHY?

#1058 

# HOW?

multi_delete_and_drag_notes

- multi selection with shift key
- multi delete
- multi drag and drop

I made it by giving state to `NoteList`

I changed lots of codes so that there would be mistakes.. If so, please let me know then I will fix it.

### CODE REVIEW

**Comment by kazup01** [comment]
I crucially wanted this feature. Thank you for your great contribution every time!
We will check it!

**Comment by tatoosh** [comment]
Please commit

**Comment by voidsatisfaction** [comment]
Not yet(want to implement)

- Key binding
  - shift: select multiple notes at once
  - ctrl(meta): select individual note
- Support multiple notes shortcut delete

**Comment by voidsatisfaction** [comment]
I implemented follows in addition to multi deletions and multi move notes to other folder

## Multi Pins to top

multi-pintotop

## Multi Select with `shift + arrow-down(up as well)` + Multi Delete with `shift + d`

multi-delete-with-arrows

I tried `Ctrl + leftClick` to multi selection. But it pops up context menu therefore I thought it is also okay to use `shift` instead

**Comment by voidsatisfaction** [comment]
@kazup01 Please review this pull request ^^

**Comment by kazup01** [comment]
Great! Thank you :)

File: browser/main/NoteList/index.js [hunk 1]
```diff
@@ -124,27 +144,49 @@ class NoteList extends React.Component {
     }
   }
 
+  focusNote (selectedNoteKeys, noteKey) {
+    let { router } = this.context
+    let { location } = this.props
```
Comment (sota1235) [inline browser/main/NoteList/index.js]: plz use `const` instead of `let` 🙏

Comment (sota1235) [reply browser/main/NoteList/index.js]: (we know there are many `let` in existing code. but it doesn't make sense, so we want to use `const` as possible as we can)

File: browser/main/NoteList/index.js [hunk 2]
```diff
@@ -124,27 +144,49 @@ class NoteList extends React.Component {
...
 
-    router.push({
-      pathname: location.pathname,
-      query: {
-        key: this.notes[targetIndex].storage + '-' + this.notes[targetIndex].key
-      }
-    })
+    if (!shiftKeyDown) { selectedNoteKeys = [] }
+    const priorNote = Object.assign({}, this.notes[targetIndex])
+    const priorNoteKey = `${priorNote.storage}-${priorNote.key}`
```
Comment (sota1235) [inline browser/main/NoteList/index.js]: You can use `getNoteKey` for creating note key.

File: browser/main/NoteList/index.js [hunk 3]
```diff
@@ -153,23 +195,31 @@ class NoteList extends React.Component {
...
     let { router } = this.context
     let { location } = this.props
+    let { selectedNoteKeys, shiftKeyDown } = this.state
 
     let targetIndex = this.getTargetIndex()
 
-    if (targetIndex === this.notes.length - 1) {
+    if (targetIndex === this.notes.length - 1 && shiftKeyDown) {
+      return
+    } else if (targetIndex === this.notes.length - 1) {
```
Comment (sota1235) [inline browser/main/NoteList/index.js]: Please add some variables to keep value of `targetIndex === this.notes.length - 1`

File: browser/main/NoteList/index.js [hunk 4]
```diff
@@ -153,23 +195,31 @@ class NoteList extends React.Component {
...
-    router.push({
-      pathname: location.pathname,
-      query: {
-        key: this.notes[targetIndex].storage + '-' + this.notes[targetIndex].key
-      }
-    })
+    if (!shiftKeyDown) { selectedNoteKeys = [] }
+    const nextNote = Object.assign({}, this.notes[targetIndex])
+    const nextNoteKey = getNoteKey(nextNote)
+    if (selectedNoteKeys.includes(nextNoteKey)) {
```
Comment (sota1235) [inline browser/main/NoteList/index.js]: same as https://github.com/BoostIO/Boostnote/pull/1070/files#r150545856

File: browser/main/NoteList/index.js [hunk 5]
```diff
@@ -186,17 +236,18 @@ class NoteList extends React.Component {
...
-      query: {
-        key: this.notes[targetIndex].storage + '-' + this.notes[targetIndex].key
-      }
-    })
+    selectedNoteKeys = []
+    const nextNote = Object.assign({}, this.notes[targetIndex])
+    const nextNoteKey = `${nextNote.storage}-${nextNote.key}`
+    selectedNoteKeys.push(nextNoteKey)
+
+    this.focusNote(selectedNoteKeys, nextNoteKey)
```
Comment (sota1235) [inline browser/main/NoteList/index.js]: This logic seems appearing three times.
Can't you add method for this logic?

File: browser/main/NoteList/index.js [hunk 6]
```diff
@@ -302,6 +363,22 @@ class NoteList extends React.Component {
   handleNoteClick (e, uniqueKey) {
     let { router } = this.context
     let { location } = this.props
+    let { shiftKeyDown, selectedNoteKeys } = this.state
+
+    if (shiftKeyDown && selectedNoteKeys.includes(uniqueKey)) {
+      const newSelectedNoteKeys = [...selectedNoteKeys].filter((noteKey) => noteKey !== uniqueKey)
```
Comment (sota1235) [inline browser/main/NoteList/index.js]: You can use `findNoteByKey()`

File: browser/main/NoteList/index.js [hunk 7]
```diff
@@ -352,16 +429,23 @@ class NoteList extends React.Component {
...
 
   handleNoteContextMenu (e, uniqueKey) {
     const { location } = this.props
-    const note = this.notes.find((note) => {
-      const noteKey = `${note.storage}-${note.key}`
-      return noteKey === uniqueKey
-    })
+    const { selectedNoteKeys } = this.state
+    const note = findNoteByKey(this.notes, uniqueKey)
+    const noteKey = `${note.storage}-${note.key}`
```
Comment (sota1235) [inline browser/main/NoteList/index.js]: You can use `getNoteKey`

File: browser/main/NoteList/index.js [hunk 8]
```diff
@@ -370,42 +454,101 @@ class NoteList extends React.Component {
...
+    this.setState({ selectedNoteKeys: [] })
   }
 
-  deleteNote (e, uniqueKey) {
-    this.handleNoteClick(e, uniqueKey)
-    ee.emit('detail:delete')
+  deleteNote () {
+    const { dispatch } = this.props
+    const { selectedNoteKeys } = this.state
+    const notes = this.notes.map((note) => Object.assign({}, note))
```
Comment (sota1235) [inline browser/main/NoteList/index.js]: [question] What do you mean by this code...?Can't you write simply `const notes = this.notes`?

File: browser/main/NoteList/index.js [hunk 9]
```diff
@@ -370,42 +454,101 @@ class NoteList extends React.Component {
...
+        data.forEach((item) => {
+          dispatch({
+            type: 'DELETE_NOTE',
+            storageKey: item.storageKey,
+            noteKey: item.noteKey
+          })
+        })
+      })
+      .catch((err) => {
+        console.error('Cannot Delete note: ' + err)
```
Comment (sota1235) [inline browser/main/NoteList/index.js]: I feel we should notice to user that deleting note was failed.
Can you show some modal window?
(If it is difficult, we can implement later.)

File: browser/main/SideNav/StorageItem.js [hunk 10]
```diff
@@ -142,40 +142,50 @@ class StorageItem extends React.Component {
     e.target.style.backgroundColor = e.dataTransfer.getData('defaultColor')
   }
 
+  dropNote (storage, folder, dispatch, location, noteData) {
+    noteData = noteData.filter((note) => folder.key !== note.folder)
+    const newNoteData = noteData.map((note) => Object.assign({}, note, {storage: storage, folder: folder.key}))
+    if (noteData.length === 0) return
```
Comment (sota1235) [inline browser/main/SideNav/StorageItem.js]: You can write this code on line 147

**Comment by sota1235** [comment]
btw, this feature is so useful :) Thank you so much for contributing!

**Comment by voidsatisfaction** [comment]
@sota1235 Thank you for review! I will check as soon as possible ^^

**Comment by voidsatisfaction** [comment]
@sota1235 @kazup01 I reviewed feedbacks.

**Comment by kazup01** [comment]
Thank you @voidsatisfaction :)

**Comment by kazup01** [comment]
Hi @voidsatisfaction , I'm so sorry for our long absence. Could you fix conflicts?

**Comment by voidsatisfaction** [comment]
@kazup01 sure. I will do that.

**Comment by kazup01** [comment]
Thanks @voidsatisfaction ;)

**Comment by kazup01** [comment]
Merged. Thanks @voidsatisfaction !
I will write this pull request to the v0.8.18 release note :)

**Comment by voidsatisfaction** [comment]
@kazup01 I'm really glad to hear that ^^

**Comment by thomashoddinott** [comment]
multiselect with `shift` works, but as soon as I try to drag I lose my selected notes.

**Pull request merged**